bench/workgen/latency_metric.py

```diff
@@ -0,0 +1,244 @@
+# successors. We intend this dedication to be an overt act of
+# relinquishment in perpetuity of all present and future rights to this
+# software under copyright law.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+
+# latency_metric.py
+# Print latency metrics for workgen runs that generate monitor.json
+import json, sys
+from datetime import datetime
+
+# A 'safe' divide shown as a string.
+def divide(a, b):
+    if b == 0.0:
+        if a == 0.0:
+            return '0.000  (no entries)'
+        else:
+            return 'infinite  (divide by zero)'
+    else:
+        return '%.3f' % (float(a) / float(b))
+
+def value_as_str(self):
+    return '%.3f' % self.latency_average()
+
+# A collection of statastics that are related to a specific condition
+# during the run, for example during checkpoints or not during checkpoints.
+class Digest:
+    def __init__(self):
+        self.entries = 0
+        self.ops = 0
+        self.lat = 0
+        self.lat_99 = 0
+        self.lat_99_raw = 0
+        self.lat_max = 0
+        self.secs = 0.0
+
+    def entry(self, secs, ops, lat, lat_99, lat_max):
+        self.secs += secs
+        self.ops += ops
+        self.lat += lat * ops
+        self.lat_99_raw += lat_99
+        self.lat_99 += lat_99 * ops
+        if lat_max > self.lat_max:
+            self.lat_max = lat_max
+        self.entries += 1
+
+    def time_secs(self):
+        return self.secs
+
+    def latency_99_raw_average(self):
+        return float(self.lat_99_raw) / float(self.entries)
+
+    def latency_max(self):
+        return self.lat_max
+
+    def latency_average(self):
+        return float(self.lat) / float(self.ops)
+
+    def dump(self, prefix):
+        print(prefix + 'json entries: ' + str(self.entries))
+        print(prefix + 'operations: ' + str(self.ops))
+        print(prefix + 'total latency us: ' + str(self.lat))
+        print(prefix + 'latency 99% us, weighted sum: ' + str(self.lat_99))
+        print(prefix + 'latency 99% us, raw sum: ' + str(self.lat_99_raw))
+        print(prefix + 'latency max us: ' + str(self.lat_max))
+        print(prefix + 'elapsed secs: ' + str(self.secs))
+
+class Metric:
+    def __init__(self, name, desc):
+        self.name = name
+        self.desc = desc
+        self.value = 0.0
+
+    def set_value(self, value):
+        self.value = value
+
+# A set of latency metrics collected for a single run.
+class FileMetrics:
+    def __init__(self, filename):
+        self.filename = filename
+        all = []
+        self.latency_avg = m = Metric('Average latency reads us',
```

Comment:
I think this is a great starting place. Looking at the sample output in the ticket, some of the descriptions are fairly brief. That's fine. But it would be good if the comments in this function compensated for that to make sure readers understand what they're looking at.